From f4410b2cbfbe88835cac75f162af1711afad9537 Mon Sep 17 00:00:00 2001
From: Jonas Johansson <jonas.johansson@westermo.se>
Date: Thu, 7 Sep 2017 14:42:27 +0200
Subject: [PATCH 24/38] route/link: add support for assigning STP group per
 VLAN
Organization: Westermo Network Technologies AB

API:
	rtnl_link_bridge_vlan_set_sid()
	rtnl_link_bridge_vlan_get_sid()

Signed-off-by: Jonas Johansson <jonas.johansson@westermo.se>
Signed-off-by: Joachim Nilsson <troglobit@gmail.com>
---
 include/linux-private/linux/if_bridge.h |  1 +
 include/netlink/route/link/bridge.h     |  4 ++++
 lib/route/link/bridge.c                 | 28 +++++++++++++++++++++++++
 libnl-route-3.sym                       |  2 ++
 4 files changed, 35 insertions(+)

diff --git a/include/linux-private/linux/if_bridge.h b/include/linux-private/linux/if_bridge.h
index 2e80dc2..1192545 100644
--- a/include/linux-private/linux/if_bridge.h
+++ b/include/linux-private/linux/if_bridge.h
@@ -122,6 +122,7 @@ enum {
 	IFLA_BRIDGE_MODE,
 	IFLA_BRIDGE_VLAN_INFO,
 	IFLA_BRIDGE_VLAN_TUNNEL_INFO,
+	IFLA_BRIDGE_VLAN_STP_INFO,
 	__IFLA_BRIDGE_MAX,
 };
 #define IFLA_BRIDGE_MAX (__IFLA_BRIDGE_MAX - 1)
diff --git a/include/netlink/route/link/bridge.h b/include/netlink/route/link/bridge.h
index ffd2354..01542d3 100644
--- a/include/netlink/route/link/bridge.h
+++ b/include/netlink/route/link/bridge.h
@@ -29,6 +29,7 @@ struct rtnl_link_bridge_vlan
 	uint16_t                pvid;
 	uint32_t                vlan_bitmap[RTNL_LINK_BRIDGE_VLAN_BITMAP_LEN];
 	uint32_t                untagged_bitmap[RTNL_LINK_BRIDGE_VLAN_BITMAP_LEN];
+        uint32_t                sid[RTNL_LINK_BRIDGE_VLAN_BITMAP_MAX];
 };
 
 /**
@@ -76,6 +77,9 @@ extern int	rtnl_link_bridge_vlan_foreach(struct rtnl_link *link,
 					      int (*cb)(struct rtnl_link *,
 							const struct bridge_vlan_info *, void *),
 					      void *arg);
+extern int	rtnl_link_bridge_vlan_set_sid(struct rtnl_link *link, unsigned int vid,
+					      unsigned int sid);
+extern int      rtnl_link_bridge_vlan_get_sid(struct rtnl_link *link, unsigned int vid);
 
 extern int	rtnl_link_bridge_unset_flags(struct rtnl_link *, unsigned int);
 extern int	rtnl_link_bridge_set_flags(struct rtnl_link *, unsigned int);
diff --git a/lib/route/link/bridge.c b/lib/route/link/bridge.c
index fe7272d..130346f 100644
--- a/lib/route/link/bridge.c
+++ b/lib/route/link/bridge.c
@@ -447,6 +447,8 @@ static struct nla_policy br_af_attrs_policy[IFLA_BRIDGE_MAX+1] = {
 	[IFLA_BRIDGE_FLAGS]		= { .type = NLA_U16 },
 	[IFLA_BRIDGE_MODE]		= { .type = NLA_U16 },
 	[IFLA_BRIDGE_VLAN_INFO]		= { .type = NLA_UNSPEC },
+	[IFLA_BRIDGE_VLAN_TUNNEL_INFO]	= { .type = NLA_UNSPEC },
+	[IFLA_BRIDGE_VLAN_STP_INFO]	= { .type = NLA_U16 },
 };
 
 static int bridge_parse_af(struct rtnl_link *link, struct nlattr *af_spec,
@@ -467,6 +469,9 @@ static int bridge_parse_af(struct rtnl_link *link, struct nlattr *af_spec,
 	if (attrs[IFLA_BRIDGE_MODE])
 		bd->b_mode = nla_get_u16(attrs[IFLA_BRIDGE_MODE]);
 
+	if (attrs[IFLA_BRIDGE_VLAN_STP_INFO])
+		bd->b_mode = nla_get_u16(attrs[IFLA_BRIDGE_VLAN_STP_INFO]);
+
 	rtnl_link_bridge_vlan_flush(link);
 	err = __parse_vlans(af_spec, bd);
 	if (err)
@@ -495,6 +500,9 @@ static int __fill_vlan(struct rtnl_link_bridge_vlan *bv, struct nl_msg *msg,
 	/*        (vlan.flags & BRIDGE_VLAN_INFO_RANGE_BEGIN) ? "range-begin" : "", */
 	/*        (vlan.flags & BRIDGE_VLAN_INFO_RANGE_END) ? "range-end" : ""); */
 	       
+	if (bv->sid[vid])
+		nla_put_u16(msg, IFLA_BRIDGE_VLAN_STP_INFO, bv->sid[vid]);
+
 	return nla_put(msg, IFLA_BRIDGE_VLAN_INFO, sizeof(vlan), &vlan);
 }
 
@@ -1035,6 +1043,26 @@ int rtnl_link_bridge_vlan_foreach(struct rtnl_link *link,
 	return 0;
 }
 
+int rtnl_link_bridge_vlan_set_sid(struct rtnl_link *link, unsigned int vid, unsigned int sid)
+{
+	struct bridge_data *bd = bridge_data(link);
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd->vlan_info.sid[vid] = sid;
+
+	return 0;
+}
+
+int rtnl_link_bridge_vlan_get_sid(struct rtnl_link *link, unsigned int vid)
+{
+	struct bridge_data *bd = bridge_data(link);
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	return bd->vlan_info.sid[vid];
+}
+
 /**
  * Unset flags
  * @arg link		Link object of type bridge
diff --git a/libnl-route-3.sym b/libnl-route-3.sym
index bec3f80..0831a8f 100644
--- a/libnl-route-3.sym
+++ b/libnl-route-3.sym
@@ -1246,4 +1246,6 @@ global:
 	rtnl_nat_get_action;
 	rtnl_neigh_get_entry_state;
 	rtnl_neigh_get_updated;
+	rtnl_link_bridge_vlan_set_sid;
+	rtnl_link_bridge_vlan_get_sid;
 } libnl_3_4;
-- 
2.25.1

